File: migrate_suppliers.py
#!/usr/bin/env python3
"""
Script per migrare i fornitori dal campo 'supplier' degli ingredienti
alla nuova tabella 'suppliers', eliminando i duplicati.
"""

import mysql.connector
import os
from nanoid import generate

# Connessione al database
conn = mysql.connector.connect(
    host=os.environ.get('DATABASE_HOST', 'gateway01.us-east-1.prod.aws.tidbcloud.com'),
    port=int(os.environ.get('DATABASE_PORT', 4000)),
    user=os.environ.get('DATABASE_USER', '2eCfxjqzjmDzEXM.root'),
    password=os.environ.get('DATABASE_PASSWORD', 'uNkWiWnOLqVFYIHK'),
    database=os.environ.get('DATABASE_NAME', 'kitchen_management'),
    ssl_ca='/etc/ssl/certs/ca-certificates.crt'
)

cursor = conn.cursor()

# 1. Estrai tutti i fornitori unici dal campo supplier degli ingredienti
print("Estraendo fornitori unici dagli ingredienti...")
cursor.execute("""
    SELECT DISTINCT supplier
    FROM ingredients
    WHERE supplier IS NOT NULL AND supplier != ''
    ORDER BY supplier
""")

unique_suppliers = [row[0] for row in cursor.fetchall()]
print(f"Trovati {len(unique_suppliers)} fornitori unici")

# 2. Inserisci i fornitori nella tabella suppliers
print("\nInserendo fornitori nella tabella suppliers...")
supplier_map = {}  # nome fornitore -> id

for supplier_name in unique_suppliers:
    supplier_id = generate(size=21)
    try:
        cursor.execute("""
            INSERT INTO suppliers (id, name, contact, email, phone, address, notes, createdAt, updatedAt)
            VALUES (%s, %s, NULL, NULL, NULL, NULL, NULL, NOW(), NOW())
        """, (supplier_id, supplier_name))
        supplier_map[supplier_name] = supplier_id
        print(f"  ✓ {supplier_name} -> {supplier_id}")
    except mysql.connector.IntegrityError as e:
        if "Duplicate entry" in str(e):
            # Fornitore già esistente, recupera l'ID
            cursor.execute("SELECT id FROM suppliers WHERE name = %s", (supplier_name,))
            existing_id = cursor.fetchone()[0]
            supplier_map[supplier_name] = existing_id
            print(f"  ⚠ {supplier_name} già esistente -> {existing_id}")
        else:
            raise

conn.commit()
print(f"\n✓ Migrazione completata: {len(supplier_map)} fornitori nella tabella suppliers")

# 3. Mostra statistiche
cursor.execute("SELECT COUNT(*) FROM suppliers")
total_suppliers = cursor.fetchone()[0]
print(f"\nStatistiche finali:")
print(f"  Fornitori totali: {total_suppliers}")

cursor.close()
conn.close()

File: migrate_suppliers.mjs
#!/usr/bin/env node
/**
 * Script per eseguire la migrazione fornitori da ingredienti
 * usando la procedura tRPC suppliers.migrateFromIngredients
 */

import { drizzle } from "drizzle-orm/mysql2";
import { suppliers, ingredients } from "./drizzle/schema.ts";
import { nanoid } from "nanoid";

const db = drizzle(process.env.DATABASE_URL);

console.log("🔄 Inizio migrazione fornitori...\n");

// 1. Estrai fornitori unici dagli ingredienti
const allIngredients = await db.select().from(ingredients);
const uniqueSuppliers = Array.from(
  new Set(allIngredients.map(i => i.supplier).filter(Boolean))
);

console.log(`📊 Trovati ${uniqueSuppliers.length} fornitori unici`);
console.log(`📦 Totale ingredienti: ${allIngredients.length}\n`);

// 2. Inserisci i fornitori nella tabella suppliers
let created = 0;
let skipped = 0;

for (const supplierName of uniqueSuppliers) {
  try {
    await db.insert(suppliers).values({
      id: nanoid(),
      name: supplierName,
      contact: null,
      email: null,
      phone: null,
      address: null,
      notes: null,
    });
    created++;
    console.log(`✅ Creato: ${supplierName}`);
  } catch (error) {
    if (error.message?.includes("Duplicate")) {
      skipped++;
      console.log(`⚠️  Già esistente: ${supplierName}`);
    } else {
      console.error(`❌ Errore per ${supplierName}:`, error.message);
    }
  }
}

console.log(`\n✨ Migrazione completata!`);
console.log(`   Creati: ${created}`);
console.log(`   Saltati (già esistenti): ${skipped}`);
console.log(`   Totale: ${uniqueSuppliers.length}`);

process.exit(0);
